### PR TITLE
8297692: Avoid sending per-region GCPhaseParallel JFR events in G1ScanCollectionSetRegionClosure

### PR DESCRIPTION
**Pull request by tschatzl**
Hi all,

  please review this change that significantly reduces the amount of "Code Roots" and "Optional Roots" JFR events to reduce default recording sizes significantly.

E.g. a 10min BigRamTester run creates a 23MB recording without this change, with like hundreds of thousands of these events (#gcs * #gc threads * #regions in collection set). With this change, the recording is reduced to 4MB (#gcs * #gc threads)

Testing: gha, tier1-3

Thanks,
  Thomas

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8297692](https://bugs.openjdk.org/browse/JDK-8297692): Avoid sending per-region GCPhaseParallel JFR events in G1ScanCollectionSetRegionClosure (**Enhancement** - P4)

### Reviewers
 * [Ivan Walulya](https://openjdk.org/census#iwalulya) (@walulyai - **Reviewer**)
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**)
 * [Sangheon Kim](https://openjdk.org/census#sangheki) (@sangheon - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21984/head:pull/21984` \
`$ git checkout pull/21984`

Update a local copy of the PR: \
`$ git checkout pull/21984` \
`$ git pull https://git.openjdk.org/jdk.git pull/21984/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21984`

View PR using the GUI difftool: \
`$ git pr show -t 21984`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21984.diff">https://git.openjdk.org/jdk/pull/21984.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21984#issuecomment-2465276011)
</details>

### CODE REVIEW

File: src/hotspot/share/gc/g1/g1RemSet.cpp [hunk 1]
```diff
@@ -786,92 +830,58 @@ class G1ScanCollectionSetRegionClosure : public G1HeapRegionClosure {
...
-    _code_root_scan_time(),
-    _code_trim_partially_time(),
-    _rem_set_opt_root_scan_time(),
-    _rem_set_opt_trim_partially_time() { }
-
-  bool do_heap_region(G1HeapRegion* r) {
-    // The individual references for the optional remembered set are per-worker, so we
-    // always need to scan them.
+    _opt_refs_memory_used(0) { }
+
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1RemSet.cpp:842]:
```suggestion
    _opt_refs_memory_used(0) { }

```

**Comment by tschatzl** [comment]
Fwiw, I went with splitting code root scan and optional root scan into two iterations that are each bracketed by a single per-thread JFR event now.
This also allowed a minor optimization: in the initial evacuation there can be no optional roots, so that iteration over all regions can be skipped.

**Review by walulyai — APPROVED** [review]
LGTM!

**Review by albertnetymk — APPROVED** [review]

**Review by sangheon — APPROVED** [review]
LGTM

**Comment by tschatzl** [comment]
Thanks @sangheon @walulyai @albertnetymk for your reviews

/integrate

**Pull request closed without merging**